src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs

```diff
@@ -446,34 +468,65 @@ private bool ParametersAreViable(MethodInfo mi)
             return true;
         }
 
-        private bool ParametersAreExact(MethodInfo mi)
+        private int CountOfExactParameterMatches(MethodInfo mi)
         {
+            int exactMatches = 0;
+
             // We can already assume that the counts match by this point
             var parameters = mi.GetParameters();
 
             for (var i = 0; i < parameters.Length; i++)
             {
-                var candidateParameter = parameters[i];
-
-                var parameterType = candidateParameter.ParameterType;
-
-                var actualArgumentType = GetExpectedParameterTypeByIndex(i);
-
-                if (actualArgumentType == null)
+                if (ParameterIsExact(parameters, i) == true)
                 {
-                    // Skip the rest of this check, as we can't know the type
-                    continue;
+                    exactMatches++;
                 }
+            }
+
+            return exactMatches;
+        }
 
-                if (parameterType != actualArgumentType)
+        private bool ParametersAreExact(MethodInfo mi)
+        {
+            // We can already assume that the counts match by this point
+            var parameters = mi.GetParameters();
+
+            for (var i = 0; i < parameters.Length; i++)
+            {
+                if (ParameterIsExact(parameters, i) != true)
                 {
+                    // This will return false on NULL or False.
                     return false;
                 }
             }
 
             return true;
         }
 
+        /// <summary>
+        /// True is exact match.
+        /// NULL if the passed parameter is null, which stands for unknown.
+        /// False if not an exact match.
+        /// </summary>
+        /// <param name="parameters">The parameters of a candidate method.</param>
+        /// <param name="index">The index of the parameters being inspected.</param>
+        /// <returns>Nullable bool stating whether this is a known exact match.</returns>
+        private bool? ParameterIsExact(ParameterInfo[] parameters, int index)
+        {
+            var candidateParameter = parameters[index];
+
+            var parameterType = candidateParameter.ParameterType;
+
+            var actualArgumentType = GetExpectedParameterTypeByIndex(index);
+
+            if (actualArgumentType == null)
+            {
+                return null;
+            }
+
+            return parameterType == actualArgumentType;
```

Comment:
Why is the parameter match trying to be exact vs. IsAssignableFrom?

Reply:
There might be many methods on a type where the type is assignable:
```
void Method(int i);
void Method(object i);
void Method<T>(T i);
void Method(long i);
```

Reply:
What we really need here is "closest assignable match"

Reply:
But, we're already in a very dangerous state if we have not matched on mdToken